### ISSUE TITLE
Deleting the last BMH using a credentials secret, causes the secret to be also deleted.

### ISSUE THREAD

**Issue by raffaelespazzoli** [metal3-io/baremetal-operator#1347]
**What steps did you take and what happened:**
Deleting the last BMH using a credentials secret, causes the secret to be also deleted.
This is not desirable.

**What did you expect to happen:**
The secret to be left alone.

**Anything else you would like to add:**
I don't understand what was the intention with using the owner reference in this case. It does seem unnecessary.

**Environment:**

- Baremetal Operator version: OCP 4.13
- Environment (metal3-dev-env or other):

/kind bug

**Comment by zaneb** [comment]
I think the logic was that we should not leave credentials lying around when they are not used by anything.
Some more information about why it's undesirable would be helpful.

**Comment by zaneb** [comment]
We also user the owner reference as a reverse link from the Secret to the BMH that owns it, so we can easily tell which BMH(s) need to be reconciled when a Secret is modified without having to search all of them. controller-runtime makes this very simple to set up by just calling `Owns(&corev1.Secret{})` on the controller manager, so that may have been part of the reason it was used initially.

**Comment by raffaelespazzoli** [comment]
the ownership concept in kubernetes generally refers to the fact a controller of resource A has created resource B in response to resource A configuration. So, in a way resource B exists only for the fact that resource A exists and therefore A owns B. The garbage collection logic aligns with this assumption and will delete B on behalf of controller A if A is deleted. This is to avoid writing trivial finalizers.
This is not the case for this BMH controller, you don't create the secret therefore you don't own it. I get that you get some convenient behavior by setting the owner, but what you are doing is conceptually wrong and it is hurting situations where the secret should not be deleted.

**Comment by lentzi90** [comment]
I agree that the current behavior is surprising. Perhaps a global flag would be cleanest?

**Comment by Rozzii** [comment]
So is there any conclusion as far as I can understand the closest to an acceptable compromise would be to turn this feature into a configurable one right ? @dtantsur @s3rj1k @lentzi90 @raffaelespazzoli @zaneb  ?

**Comment by hardys** [comment]
This was also [recently discussed](https://github.com/metal3-io/baremetal-operator/pull/1380#discussion_r1347372523) in relation to networkData secrets - it seems we're not consistently setting ownership of those (and presumably meta/userData) so we might want to consider if there's a common (but configurable) way to solve this?

**Comment by zaneb** [comment]
We do consistently set ownership of user/meta/networkData and BMC Secrets.
We don't set ownership of preprovisioningNetworkData unless we're actually using it, so that's inconsistent.

**Comment by zaneb** [comment]
> We do consistently set ownership of user/meta/networkData and BMC Secrets. We don't set ownership of preprovisioningNetworkData unless we're actually using it, so that's inconsistent.

Actually, scratch that. We do _not_ set ownership of user/meta/networkData because it's not required that we trigger a reconcile when these change.

**Comment by zaneb** [comment]
> So is there any conclusion as far as I can understand the closest to an acceptable compromise would be to turn this feature into a configurable one right ?

TBH I see this as an Open Source antipattern - the developers, who know and understand the software the best, can't agree so they punt the decision to the person least equipped to make it (in this case the operator of the metal³ deployment).

How can we ask the operator to decide at deployment time between timely responses to changes and keeping unreferenced Secrets around? How would they decide if we can't?

Any end user who wants to keep their unreferenced Secrets around is free to add another owner reference to the Secret that will prevent it being deleted. To my mind that's better than an annotation on the BMH because you can see what it does directly, and there's no bizarre side-effects like changes to the Secret being ignored until they suddenly take effect at unexpected times.

**Comment by lentzi90** [comment]
> Any end user who wants to keep their unreferenced Secrets around is free to add another owner reference to the Secret that will prevent it being deleted. To my mind that's better than an annotation on the BMH because you can see what it does directly, and there's no bizarre side-effects like changes to the Secret being ignored until they suddenly take effect at unexpected times.

I would rather treat the current behavior as a bug and just change it. Metal3 is not the owner of these secrets and should therefore not put owner references on them. Finalizers on the other hand could make sense to prevent deletion while in use.

**Comment by Rozzii** [comment]
I think it is clear that this is an issue regardless of what will be the solution.
/kind bug
/triage accepted

**Comment by Rozzii** [comment]
This is a serious problem, caused quite a big issue in https://github.com/metal3-io/baremetal-operator/issues/3304  , but I don't think we will have a fix for 1.14 . For me this is priority so I will aim this for 1.15
